docs/architecture/adr-056-proving-amnesia-attacks.md

```diff
@@ -0,0 +1,120 @@
+
+### Amnesia Attack
+
+The currently only known form of global evidence stems from [flip flopping](https://github.com/tendermint/spec/blob/master/spec/consensus/light-client/accountability.md#flip-flopping) attacks. The schematic below explains one scenario where an amnesia attack, a form of flip flopping, can occur such that two sets of honest nodes, C1 and C2, commit different blocks.
+
+![](../imgs/tm-amnesia-attack.png)
+
+1. C1 and F send PREVOTE messages for block A.
+2. C1 sends PRECOMMIT for round 1 for block A.
+3. A new round is started, C2 and F send PREVOTE messages for a different block B.
+4. C2 and F then send PRECOMMIT messages for block B.
+5. F breaks the lock and goes back and sends PRECOMMIT messages in round 1 for block A.
+
+
+This creates a fork on the main chain.  Back to the past, another form of flip flopping, creates a light fork (capable of fooling those not involved in consensus), in a similar way, with F taking the precommits from C1 and forging a commit from them.
+
+## Decision
+
+As the distinction between these two attacks (amnesia and back to the past) can only be distinguished by confirming with all validators (to see if it is a full fork or a light fork), for the purpose of simplicity, these attacks will be treated as the same.
+
+Currently, the evidence reactor is used to simply broadcast and store evidence. Instead of perhaps creating a new reactor for the specific task of verifying these attacks, the current evidence reactor will be extended.
+
+The process begins with a light client receiving conflicting headers (in the future this could also be a full node during fast sync), which it sends to a full node to analyse. As part of [evidence handling](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-047-handling-evidence-from-light-client.md), this could be deduced into potential amnesia evidence
+
+```golang
+type PotentialAmnesiaEvidence struct {
+	V1 []*types.Vote
+	V2 []*types.Vote
+
+	timestamp time.Time
+}
+```
+
+*NOTE: Unlike prior evidence types, `PotentialAmnesiaEvidence` and `AmnesiaEvidence` are processed as a batch instead
+ of individually. This will require changes to much of the API.*
+
+ *NOTE: `PotentialAmnesiaEvidence` could be constructed for when 1/3 or less vote in two different rounds but as it is not currently detected nor can it cause a fork, it will be ignored.*
+
+The evidence should contain the precommit votes for the intersection of validators that voted for both rounds. The votes should be all valid and the height and time that the infringement was made should be within:
+
+`Trusting period - Amnesia trial period - Gossip safety margin`
```

Comment:
Trusting period = UnbodingPeriod - Amnesia trial period - Gossip safety margin
Evidence.MaxAgeDuration = UnbondingPeriod + Amnesia trial period + Gossip safety margin

?

Reply:
I thought the trusting period was the unbonding period?

Reply:
Basically we want to check before we send potential amnesia evidence, that it will reach everyone and that we can wait out the entire trial period and then commit the evidence whilst the perpetrator is still bonded

Reply:
Hence the sending node of potential evidence should check that time.Now() < ev.Time() + Unbonding Period - Trial Period - Gossip Margin

Reply:
> I thought the trusting period was the unbonding period?

trusting period is unbonding period minus evidence propagation delay + delay for punishment.

Reply:
> Unbonding Period - Trial Period - Gossip Margin

that's trusting period essentially

Reply:
> > Unbonding Period - Trial Period - Gossip Margin
> 
> that's trusting period essentially

but without the trial period

Reply:
so that means I can take out the gossip margin and just have the duration as trust period - trial period

Reply:
> but without the trial period

no, I think it should include trial period as well.

> so that means I can take out the gossip margin and just have the duration as trust period - trial period

problem is there's no notion of trusting period in tendermint (except for lite2 where it's defined as config option)

Reply:
okay so MaxEvidenceAge - Trial Period

Reply:
no:) MaxEvidenceAge - Amnesia Trial - Gossip Margin

Reply:
but instead of adding new consensus params, we can simply say MaxEvidenceAge should be UnbodingPeriod - Amnesia Trial - Gossip Margin

Reply:
But we do need to at least define Trial Period, either as a consensus param or a constant

Reply:
It could be half the max evidence age say but nodes need to know how long they can give other nodes to respond before they propose amnesia evidence

Reply:
okay. let's make it a consensus parameter (1 day is good imo)

Reply:
Done